### DISCUSSION TITLE
How to develop I2C device directly on AI-deck

### DISCUSSION THREAD

**Discussion by mfxjx000205** [bitcraze/discussions#1172] [Q&A] [answered]
Hi, Bitcraze,

In the past few weeks of development, I've run into a very tricky problem. Since the Crazyflie 2.1 flight control board is weak in terms of arithmetic and memory, our work hopes to make some of the expansion boards with sensors independent. **Driven only by the ai-deck without going through crazyflie.** 

Based on the above idea, we did the following:

1. As this [page](https://github.com/ETH-PBL/Matrix_ToF_Drones) shows, we referenced ETH-Zurich's github to make an identical piece of **tof-deck.** The tof-deck is a ranging expansion board consisting of two vl53l5cx sensors, and a TCA6408 I2C expander, capable of communicating with other devices via an I2C interface. Colleagues who are familiar with multiranger-deck should not need much introduction.

2. We used the driver code from the page above and the tof-deck was driven quite correctly on crazyflie. So, we wanted to drive the tof-deck directly, via I2C, on the AI-deck, and that's where we hit a roadblock.

  a) We connect the I2C interface of the tof-deck, to the I2C interface of the AI-deck, SDA to pin R8 (SPARE_I2C_SDA), SCL to pin R9(SPARE_I2C_SCL).

  b) We refer to the I2C example [function](https://github.com/GreenWaves-Technologies/gap_sdk/blob/master/examples/gap8/basic/interfaces/i2c/i2c_bmp280/test_i2c_bmp280.c) in gap_sdk and want to test whether AI-deck can directly drive tof-deck. The expected way to test this is to modify the value of a register on a TCA6408.

  c) The results of the test are unsatisfactory; no matter what method we use, the bus always reads 255, which is 11111111. According to the operating principle of I2C, the data bus is always high when it is not started. This should be the bus not starting.

3. We want to ask:

  a) In your opinion, can the pin8 and pin9 reserved on top of the AI-deck be used to drive other I2C devices independently. In other words, **is it feasible to use the two SPARE_I2C pins directly from the AI-deck's schematic, and do I need to do some additional configurations such as GPIO's?**

  b) Are there any special changes made to the AI-deck compared to other development boards for the GAP8 that prevent I2C devices from being driven **directly and correctly**? Should I make some changes in the bootloader of the AI-deck?

  c) Regarding this need of ours, we would like to ask Bitcraze if they have any further development suggestions.

Best Wishes!

Hanjie.

**Comment by tobbeanton** [comment]
In theory this is a multi-master I2C setup which should work if the software handles it correctly. Since I have not tested this myself it is hart to say whats wrong, perhaps @hmllr has some ideas? Would it be possible to contact the authors at ETHZ for some help?

**Reply by hmllr** [reply, marked as answer]
I also think this should theoretically work, I think we never tried it on the AI-deck/GAP8, but we once had a student who ported the ToF driver to a Mr.Wolf SoC (which is very similar to GAP8) and another one to GAP9 (which is also not too far off), so I don't see any problem with it. Maybe you are missing the pull-up resistors for the I2C? As they are usually on the Crazyflie itself, they are not on the AI-deck? Other than that it should be pretty straight forward, the I2C is the standard function of those pins. Maybe checking what is actually happening on the bus with an oscilloscope or logic analyzer can help as well. Good luck,
Hanna
